core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1322,27 +1349,46 @@ public static void writeOperationNoSize(BufferedChecksumStreamOutput out, Transl
+     *
+     * @param generation the current translog generation
+     * @throws IOException if an I/O exception occurred during any file operations
+     */
+    void foldGeneration(final long generation) throws IOException {
+        assert writeLock.isHeldByCurrentThread();
+        try {
+            final TranslogReader reader = current.closeIntoReader();
+            readers.add(reader);
+            final Path checkpoint = location.resolve(CHECKPOINT_FILE_NAME);
+            assert Checkpoint.read(checkpoint).generation == generation;
+            final Path generationCheckpoint =
+                    location.resolve(getCommitCheckpointFileName(generation));
+            Files.copy(checkpoint, generationCheckpoint);
+            IOUtils.fsync(generationCheckpoint, false);
+            IOUtils.fsync(generationCheckpoint.getParent(), true);
+            // create a new translog file; this will sync it and update the checkpoint data;
+            current = createWriter(generation + 1);
+            logger.trace("current translog set to [{}]", current.getGeneration());
+        } catch (final Exception e) {
+            IOUtils.closeWhileHandlingException(this); // tragic event
+            throw e;
+        }
+    }
+
     @Override
     public long prepareCommit() throws IOException {
-        try (ReleasableLock lock = writeLock.acquire()) {
+        try (ReleasableLock ignored = writeLock.acquire()) {
             ensureOpen();
             if (currentCommittingGeneration != NOT_SET_GENERATION) {
-                throw new IllegalStateException("already committing a translog with generation: " + currentCommittingGeneration);
+                throw new IllegalStateException("already committing a translog with generation: " +
+                        currentCommittingGeneration);
             }
-            currentCommittingGeneration = current.getGeneration();
-            TranslogReader currentCommittingTranslog = current.closeIntoReader();
-            readers.add(currentCommittingTranslog);
-            Path checkpoint = location.resolve(CHECKPOINT_FILE_NAME);
-            assert Checkpoint.read(checkpoint).generation == currentCommittingTranslog.getGeneration();
-            Path commitCheckpoint = location.resolve(getCommitCheckpointFileName(currentCommittingTranslog.getGeneration()));
-            Files.copy(checkpoint, commitCheckpoint);
-            IOUtils.fsync(commitCheckpoint, false);
-            IOUtils.fsync(commitCheckpoint.getParent(), true);
-            // create a new translog file - this will sync it and update the checkpoint data;
-            current = createWriter(current.getGeneration() + 1);
-            logger.trace("current translog set to [{}]", current.getGeneration());
-
-        } catch (Exception e) {
+            final long generation = current.getGeneration();
+            currentCommittingGeneration = generation;
+            foldGeneration(generation);
+        } catch (final Exception e) {
```

Comment:
Is this catch block still needed? `foldGeneration` has one just like it. Probably worth adding a comment if it is.